src/util/json_stream.h

```diff
@@ -0,0 +1,151 @@
+/*******************************************************************\
+
+Module:
+
+Author: Peter Schrammel
+
+\*******************************************************************/
+
+
+#ifndef CPROVER_UTIL_JSON_STREAM_H
+#define CPROVER_UTIL_JSON_STREAM_H
+
+#include "json.h"
+#include "invariant.h"
+
+class json_stream_objectt;
+class json_stream_arrayt;
+
+/// New class whose aim is to stream newly created JSON
+/// objects directly instead of waiting for the object
+/// to be fully formed in memory and then print it.
+/// 
+/// This approach was necessary to fix some issues with
+/// CBMC running out of memory when it had to print some
+/// huge traces for some programs.
+class json_streamt
+{
+public:
+  void close()
+  {
+    if(open)
+    {
+      output_current();
+      output_finalizer();
+      open = false;
+    }
+  }
+
+protected:
+  json_streamt(std::ostream &_out, unsigned _indent)
+    : open(true), out(_out), indent(_indent), first(true), current(nullptr)
+  {
+  }
+
+  bool open;
+  std::ostream &out;
+
+  unsigned indent;
+  bool first;
+
+  typedef std::map<std::string, jsont> objectt;
+  objectt object;
+
+  std::unique_ptr<json_streamt> current;
+  json_stream_arrayt &create_current_array();
+  json_stream_objectt &create_current_object();
+
+  void output_delimiter();
+
+  virtual void output_current()
+  {
+    // no-op
+  }
+
+  virtual void output_finalizer()
+  {
+    // no-op
+  }
+};
+
+/// Specialisation of the json_streamt class to handle
+/// JSON arrays to be output immediately in the stream.
+class json_stream_arrayt : public json_streamt
+{
+public:
+  json_stream_arrayt(std::ostream &out, unsigned indent = 0);
+
+  ~json_stream_arrayt()
+  {
+    close();
+  }
+
+  /// Push back a JSON object into the current array stream.
+  void push_back(const jsont &json)
+  {
+    PRECONDITION(open);
+    output_current();
+    output_delimiter();
+    json.output_rec(out, indent+1);
+  }
+
+  /// Push back a JSON object into the current array stream.
+  jsont &push_back()
+  {
+    PRECONDITION(open);
+    output_current();
+    return object[""];
+  }
+
+  json_stream_objectt &push_back_stream_object();
+  json_stream_arrayt &push_back_stream_array();
+
+protected:
+  void output_current() override;
+  void output_finalizer() override;
+};
+
+/// Specialisation of the json_streamt class to handle
+/// JSON objects to be output immediately in the stream.
+class json_stream_objectt:public json_streamt
+{
+public:
+  json_stream_objectt(std::ostream &out, unsigned indent = 0);
+
+  /// Provide key-value lookup capabilities for the
+  /// JSON object.
+  /// \param key: The key to be looked up inside the 
+  ///   attributes of the JSON object.
+  jsont &operator[](const std::string &key)
```

Comment:
How does this fit in a "stream" class?!

Reply:
It allows mixing streamed and non-streamed json elements. The API enforces that the parent of a streamed element must always be a streamed element.

Reply:
Why does it even make sense to mix them? Or is my above thought that a `json_streamt` is actually also some form of `jsont` that actually holds data?!